### PR TITLE
Add Identity-Aware Proxy Verification Sidecar

### PR DESCRIPTION
**Pull request by tariqajyusuf**
Add a sidecar to pull and update the state of Identity-Aware Proxy on a App Engine Flex application. This runs an automatic updater that retrieves IAP metadata and verification keys and puts it out to a specific file. These files will be used to automatically run verification as a defense-in-depth measure against attackers that could access the VM directly.

### CODE REVIEW

**Comment by tariqajyusuf** [comment]
I'm playing around with `docker_install_deps.sh` to see if I can get the test setup working. The rest of the PR is done.

**Comment by tariqajyusuf** [comment]
There we go, that fixed it. Tests are passing.

**Comment by karan** [comment]
+R @Cynocracy

File: iap_watcher/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+RUN mkdir /iap_watcher
+VOLUME /iap_watcher
+
+RUN mkdir -p /home/vmagent/iap_watcher
+WORKDIR /home/vmagent/iap_watcher
+ADD iap_watcher.py .
+ENTRYPOINT ["./iap_watcher.py", "--key=AEF_IAP_state", "--output_state_file=/iap_watcher/state.json"]
+
+RUN echo "15 */12 * * * curl "https://www.gstatic.com/iap/verify/public_key" > /iap_watcher/iap_verify_keys.txt" >> .tmp_cron
```
Comment (kmbarry1) [inline iap_watcher/Dockerfile]: Could we change this to curl the JWK keys? They are easier to parse: https://www.gstatic.com/iap/verify/public_key-jwk

Comment (kmbarry1) [inline iap_watcher/Dockerfile]: Also, dumb question, will this run immediately, or will it wait 12 hours before the first fetch?

File: iap_watcher/iap_watcher.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  Args:
+    value: unicode string, the value of the metadata attribute.
+    output_file: ascii string, where to write the resulting value to.
+  """
+  if value:
+    logging.info('The latest value is %s.', value)
+    metadata = json.loads(value)
+    if metadata['enabled']:
+      f = open(output_file, 'w')
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher.py]: with open(output_file, 'w') as file:
  pass

File: iap_watcher/iap_watcher.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import signal
+
+from google_compute_engine import metadata_watcher
+
+
+def _ExitWithExceptionHandle(signum, frame):
+  logging.error('Timeout when retrieving metadata. Retrying...')
+
+
+def _RetryIfValueIsEmptyHandler(value, output_file):
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher.py]: method name is a little strange, maybe UpdateStateFileFromMetadataHandler?

File: iap_watcher/iap_watcher.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        output_file=argv.output_state_file),
+      metadata_key='instance/attributes/%s' % argv.key,
+      recursive=False,
+      timeout=timeout)
+    if loop_watcher:
+      break
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser(description='Watches for IAP state changes.')
+  parser.add_argument('--key', type=str, required=True,
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher.py]: --iap_metadata_key maybe? I prefer self-documenting but slightly longer args, ymmv

File: iap_watcher/iap_watcher_test.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import os
+import shutil
+import tempfile
+import unittest
+import iap_watcher
+from jsobject import Object
+
+
+class TestIapVerifier(unittest.TestCase):
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher_test.py]: Have you considered using a mocking framework for this?

File: iap_watcher/iap_watcher_test.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    iap_watcher.Main(Object({
+        'key': 'AEF_IAP_state',
+        'timeout': None,
+        'output_state_file': path,
+      }),
+      watcher=self.metadata_watcher_,
+      loop_watcher=True)
+    self.assertTrue(os.path.isfile(path))
+
+  def testDisabledNoEnable(self):
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher_test.py]: testDisable, for consistency with the above method

File: iap_watcher/iap_watcher_test.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    iap_watcher.Main(Object({
+        'key': 'AEF_IAP_state',
+        'timeout': None,
+        'output_state_file': path,
+      }),
+      watcher=self.metadata_watcher_,
+      loop_watcher=True)
+    self.assertFalse(os.path.isfile(path))
+
+  def testEnableDisable(self):
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher_test.py]: Should we test Disable -> Enable?

File: iap_watcher/Dockerfile (hunk 1 above)
Comment (tariqajyusuf) [inline iap_watcher/Dockerfile]: Done

Comment (tariqajyusuf) [inline iap_watcher/Dockerfile]: I'm not sure, I will explicitly run it to guarantee a good initial state.

File: iap_watcher/iap_watcher.py (hunk 3 above)
Comment (tariqajyusuf) [inline iap_watcher/iap_watcher.py]: Done.

File: iap_watcher/iap_watcher.py (hunk 2 above)
Comment (tariqajyusuf) [inline iap_watcher/iap_watcher.py]: Done.

File: iap_watcher/iap_watcher.py (hunk 4 above)
Comment (tariqajyusuf) [inline iap_watcher/iap_watcher.py]: Done.

File: iap_watcher/iap_watcher_test.py (hunk 6 above)
Comment (tariqajyusuf) [inline iap_watcher/iap_watcher_test.py]: Done.

File: iap_watcher/iap_watcher_test.py (hunk 5 above)
Comment (tariqajyusuf) [inline iap_watcher/iap_watcher_test.py]: Yes, but to get the call chain working properly is too complicated in mock.

File: iap_watcher/iap_watcher_test.py (hunk 7 above)
Comment (tariqajyusuf) [inline iap_watcher/iap_watcher_test.py]: Added.

File: iap_watcher/iap_watcher.py [hunk 8]
```diff
@@ -40,8 +40,8 @@ def _RetryIfValueIsEmptyHandler(value, output_file):
     logging.info('The latest value is %s.', value)
     metadata = json.loads(value)
     if metadata['enabled']:
-      f = open(output_file, 'w')
-      f.close()
+      with open(output_file, 'w') as file:
```
Comment (Cynocracy) [inline iap_watcher/iap_watcher.py]: actually I guess you could do

with open(output_file, 'w'):

without naming it

**Comment by kmbarry1** [comment]
This was merged in as #67 and can be deleted.

**Comment by rdcastro** [comment]
Closing as this was merged as of #67

**Pull request closed without merging**